Bento/Renderable/Renderable.swift

```diff
@@ -1,14 +1,72 @@
     func render(in view: View)
+
+    /// Produce a performance hint for Bento to better decide its view reusability strategy.
+    ///
+    /// Note that having a reusability hint **does not imply** that Bento would always only reuse views when the
+    /// reusability hint matches. In other words, you **must still ensure** your composite component view handles type
+    /// mismatches in the view hierarchy correctly and gracefully.
+    ///
+    /// For example, if you have a composite component that has a small but fixed number of combinations of child
+    /// components, you may implement this requirement to improve the performance, which otherwise would involve
+    /// time in recreating views as they are queued to go on screen.
+    ///
+    /// ```swift
+    /// struct CompositeComponent: Renderable {
+    ///     let children: [AnyRenderable]
+    ///
+    ///     func render(in view: View) {
+    ///         // Logic to recreate the view hierarchy if types and orders of components do not match
+    ///     }
+    ///
+    ///     func makeReusabilityHint(using combiner: inout ReusabilityHintCombiner) {
+    ///         children.forEach { combiner.combine($0) }
+    ///     }
+    /// }
+    ///
+    /// // NOTE: These combinations would all result in different reusability hints.
+    /// //       (Assuming there are three component types `A`, `B`, and `C`.
+    /// CompositeComponent(children: [])
+    /// CompositeComponent(children: [A()])
+    /// CompositeComponent(children: [B()])
+    /// CompositeComponent(children: [C()])
+    /// CompositeComponent(children: [A(), B()])
+    /// CompositeComponent(children: [A(), C()])
+    /// CompositeComponent(children: [B(), A()])
+    /// CompositeComponent(children: [B(), C()])
+    /// CompositeComponent(children: [C(), A()])
+    /// CompositeComponent(children: [C(), B()])
+    /// CompositeComponent(children: [A(), B(), C()])
+    /// CompositeComponent(children: [A(), C(), B()])
+    /// CompositeComponent(children: [B(), A(), C()])
+    /// CompositeComponent(children: [B(), C(), A()])
+    /// CompositeComponent(children: [C(), A(), B()])
+    /// CompositeComponent(children: [C(), B(), A()])
+    /// ```
+    ///
+    /// - important: The order of `combiner.combine(_:)` matters.
+    ///
+    /// - important: Bento always considers the component type for view reusability. So components need not combine
+    ///              its own type again.
+    ///
+    /// - note: This is an optional requirement intended for composite components that contain children components
+    ///         or dynamic view hierarchies.
+    ///
+    /// - parameters:
+    ///   - combiner: The combiner to concatenate all relevant information that affects reusability.
+    func makeReusabilityHint(using combiner: inout ReusabilityHintCombiner)
 }
 
 public extension Renderable {
+    func makeReusabilityHint(using combiner: inout ReusabilityHintCombiner) {}
```

Comment:
Does it scan it through the hierarchy? e.g `C(children: [C(children: [A()], B()])` vs C(children: [C(children: [B()], A()])?

Reply:
It doesn’t do so automatically. When you say `combiner.combine(child)`, A’s part of the call it would eventually invoke `makeReusabilityHint(using:)` on the child. So as long as you combine every child using the combiner, and that all children implement this properly, the whole hierarchy would be walked over.

This can be made less crumblesome if we have a protocol for composite component e.g. `CompositeRenderable` which is also useful for auto propagation of messages like `*LifecycleAware`. But this can be done after this PR which lays the cornerstone.